src/manager.js

```diff
@@ -482,18 +509,47 @@ AdManager.prototype.loadAds = function(element, updateCorrelator) {
   }
 
   for (var i = 0; i < ads.length; i++) {
-    var thisEl = ads[i];
-
-    if ((thisEl.getAttribute('data-ad-load-state') === 'loaded') || (thisEl.getAttribute('data-ad-load-state') === 'loading')) {
+    var thisEl = ads[i],
+      slot,
+      adUnitConfig,
+      activeSizes,
+      gptSlotSizes,
+      adUnitSizes;
+
+	if ((thisEl.getAttribute('data-ad-load-state') === 'loaded') || (thisEl.getAttribute('data-ad-load-state') === 'loading')) {
       continue;
     }
 
-    var slot = this.configureAd(thisEl);
+    slot = this.configureAd(thisEl);
 
     if (slot && slot.eagerLoad) {
       slotsToLoad.push(slot);
     }
 
+    if (this.options.amazonEnabled) {
+ 
+    /**
+     * Try to use the gpt slot.getSizes method to retrieve the active sizes given the viewport parameters inside the ad config.
+     * This method is undocumented, and could be removed. When not available, fall back to all sizes specified in the ad unit itself.
+     * This is not optimal, as sizes which cannot be displayed due to the viewport dimensions will be requested from A9. It is thus used as a fallback.
+     * See Docs here https://developers.google.com/doubleclick-gpt/reference#googletagslot
+    */
+	
+      adUnitConfig = this.adUnits.units[thisEl.dataset.adUnit];
+      adUnitSizes = this.adUnitSizes(adUnitConfig.sizes)[1];
+
+      if (typeof gptSlotSizes == 'function') {
+        gptSlotSizes = slot.getSizes();
+        activeSizes = this.adSlotSizes(gptSlotSizes);
+      } else {
+        activeSizes = adUnitSizes;
+      }
+
+      if (adUnitConfig.amazonEnabled && activeSizes && activeSizes.length) {
+        this.fetchAmazonBids(thisEl.id, activeSizes);
+      }
```

Comment:
Don't we have to tie the `refresh` call (that's happening at the bottom of `loadAds`) into the Amazon bids callback to make sure it doesn't happen until the bid returns?

Reply:
The whole point here is not to wait for bids.  This is the way index implements header bidding, and i think amazon should follow the same process. I think including this in their documentation is a way of tricking inexperienced publishers to wait for amazon bids and prioritize them above their other vendors.  Since amazon is not the majority of our demand by any stretch, we should not allow a blocking callback to delay all ads on the page to amazon's benefit.  It's a problem that has plagued us in the past, where partners claim to be async but in reality they are not.

Placing the refresh here will cause the top 970 unit to continue to be plagued with poor viewability.